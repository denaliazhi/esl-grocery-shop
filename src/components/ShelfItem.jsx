/**
 * This component renders the details for an item in shelf view.
 */
import AddItem from "./AddItem";
import NumberPicker from "./NumberPicker";
import RemoveItem from "./RemoveItem";

export default function ShelfItem({ item, handleClick }) {
  return (
    <>
      <div
        className="product shelf-item"
        id={item.id}
        data-section={item.section}
      >
        <img src={`/${item.name}.png`} alt="" width="100" />
        <p>{item.name.charAt(0).toUpperCase() + item.name.slice(1)}</p>
        <p>
          ${item.unitPrice.toFixed(2)} / {item.displayUnit}
        </p>
        {item.count === 0 ? (
          <AddItem handleClick={handleClick} />
        ) : (
          <>
            <NumberPicker value={item.count} handleClick={handleClick} />
            <RemoveItem value={item.count} handleClick={handleClick} />
          </>
        )}
      </div>
    </>
  );
}
