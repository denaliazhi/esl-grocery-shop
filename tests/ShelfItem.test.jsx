import { vi, describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";

import ShelfItem from "../src/components/ShelfItem";

vi.mock("../src/components/NumberPicker", () => ({
  default: vi.fn(() => <button>Edit</button>),
}));

vi.mock("../src/components/AddItem", () => ({
  default: vi.fn(() => <button>Add</button>),
}));

describe("Shelf item", () => {
  test("renders expected content", () => {
    const item = {
      id: "x123",
      name: "apple",
      unitPrice: 1.23,
      displayUnit: "each",
      section: "produce",
      count: 1,
    };
    const { container } = render(
      <ShelfItem item={item} handleClick={() => {}} />
    );

    expect(container).toMatchSnapshot();
  });

  test("renders 'Add to cart' button when item count = 0'", async () => {
    const item = {
      id: "x246",
      name: "banana",
      unitPrice: 2.46,
      displayUnit: "each",
      section: "produce",
      count: 0,
    };
    render(<ShelfItem item={item} handleClick={() => {}} />);
    const addButton = screen.getByRole("button", { name: /Add/ });
    expect(addButton).toBeInTheDocument();
  });

  test("renders 'Edit' button when item count > 0'", async () => {
    const item = {
      id: "x137",
      name: "orange",
      unitPrice: 1.37,
      displayUnit: "each",
      section: "produce",
      count: 9,
    };
    render(<ShelfItem item={item} handleClick={() => {}} />);
    const editButton = screen.getByRole("button", { name: /Edit/ });
    expect(editButton).toBeInTheDocument();
  });
});
