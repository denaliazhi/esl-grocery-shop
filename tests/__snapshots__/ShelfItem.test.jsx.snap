// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shelf item > renders expected content 1`] = `
<div>
  <div
    class="product shelf-item"
    data-section="produce"
    id="x123"
  >
    <img
      alt=""
      src="products/apple.png"
      width="100"
    />
    <div
      class="item-label"
    >
      <h3>
        Apple
      </h3>
      <div
        class="item-price"
      >
        <p>
          $
          1.23
           /
        </p>
        <p>
          each
        </p>
      </div>
    </div>
    <button>
      Edit
    </button>
  </div>
</div>
`;
